### ISSUE TITLE
SGD: why not update gradient inplace when using weight decay?

### ISSUE THREAD

**Issue by vycezhong** [pytorch/pytorch#69189]
I notice that when using weight decay in SGD, it will create a new tensor and bind it to the `d_p`. 

https://github.com/pytorch/pytorch/blob/e534c5efd7e7a0ed476fa1e9ec2f37ccd3a1b63d/torch/optim/_functional.py#L175

I wonder why not just update the gradient inplace? like this

```py
d_p.add_(param, alpha=weight_decay)
```

It will save GPU memory.

cc @vincentqb @jbschlosser @albanD

**Comment by albanD** [comment]
Hi,

This is because the gradient `d_p` is the same as the `.grad` field on the Parameters. We don't want to modify that inplace to avoid changing this Tensor that the user might be using later.

**Comment by vycezhong** [comment]
Thanks for your explanation.
